src/step-function.ts

```diff
@@ -466,6 +465,152 @@ export namespace $SFN {
       });
     },
   });
+
+  export interface RetryFunc {
+    <Func extends ParallelFunction<any>>(
+      retry: Retry[],
+      callback: Func
+    ): ReturnType<Func>;
+    <Func extends ParallelFunction<any>>(callback: Func): ReturnType<Func>;
+  }
+
+  /**
+   * Apply Step Function's built in Retry logic to a function.
+   *
+   * ```ts
+   * new ExpressStepFunction(this, "F", (id: string) => {
+   *   const results = $SFN.retry(
+   *     async () => { await task1(id) }
+   *   )
+   * })
+   * ```
+   *
+   * By default, all errors (`States.ALL`) will be caught and retried with the default backoff strategy:
+   * MaxAttempts: 3
+   * IntervalSeconds: 1
+   * BackoffRate: 2.0
+   *
+   * Optionally provide specific error codes and backoff strategies.
+   *
+   * ```ts
+   * new ExpressStepFunction(this, "F", (id: string) => {
+   *   const results = $SFN.retry(
+   *     [{ ErrorEquals: ["Lambda.ToManyRequestsException"], BackoffRate: 2.5, IntervalSeconds: 2, MaxAttempts: 4 }],
+   *     async () => { await task1(id) }
+   *   )
+   * })
+   * ```
+   *
+   * Caveat: `$SFN.retry` does not maintain the state of mutated variables within the callback.
+   *
+   * ```ts
+   * let a = 1;
+   * await $SFN.retry(() => { a = 2; return a * 4; })
+   * return a;
+   * ```
+   *
+   * The return value will be `1`.
+   *
+   */
+  export const retry = makeStepFunctionIntegration<"$SFN.retry", RetryFunc>(
+    "$SFN.retry",
+    {
+      asl(call, context) {
+        const [retries, body] =
+          call.args.length === 1 ? [undefined, call.args[0]] : call.args;
+
+        if (!body || !isFunctionLike(body.expr)) {
+          throw new SynthError(
+            ErrorCodes.Invalid_Input,
+            "Expected $SFN.retry callback argument to be a function."
+          );
+        }
+
+        const retryValues = processRetries(retries?.expr);
+
+        const functionStates = context.evalStmt(body.expr.body, {
+          End: true,
+          ResultPath: "$",
+        });
+
+        /**
+         * If there are no states for the function, just return an empty state, nothing to do here.
+         */
+        if (!functionStates) {
```

Comment:
What if there are no states after optimization?

Reply:
Currently the optimization never removes the `StartAt` state. In the future it may be smart enough to remove Map and Parallel states that can be empty.